### PR TITLE
yoda: provide python3Packages.yoda

### PR DESCRIPTION
**Pull request by veprbl**
###### Motivation for this change

To provide `yoda` bindings for use with `python3`

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
@GrahamcOfBorg build python2Packages.yoda python3Packages.yoda

File: pkgs/development/libraries/physics/yoda/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, python2Packages, root, makeWrapper, zlib, withRootSupport ? false }:
+{ stdenv, fetchurl, pythonPackages, root, makeWrapper, zlib, withRootSupport ? false }:
```
Comment (FRidh) [inline pkgs/development/libraries/physics/yoda/default.nix]: Please do not use `pythonPackages` anymore, unless a version set is used by the caller (so also in `all-packages.nix`).

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/physics/yoda/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/development/libraries/physics/yoda/default.nix]: You want me to revert back to `python2Packages`? I thought, the purpose of using `pythonPackages` is to show that a package supports both versions.

Comment (veprbl) [reply pkgs/development/libraries/physics/yoda/default.nix]: I will be switching `yoda` to use `python3` by default, but that, probably, can not be done until I get `rivet` version 3 working.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -24497,7 +24497,9 @@ in
 
   thepeg = callPackage ../development/libraries/physics/thepeg { };
 
-  yoda = callPackage ../development/libraries/physics/yoda { };
+  yoda = callPackage ../development/libraries/physics/yoda {
+    pythonPackages = python2Packages;
```
Comment (jonringer) [inline pkgs/top-level/all-packages.nix]: python2 is EOL, please use python3Packages if possible

File: pkgs/development/libraries/physics/yoda/default.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, python2Packages, root, makeWrapper, zlib, withRootSupport ? false }:
+{ stdenv, fetchurl, fetchpatch, pythonPackages, root, makeWrapper, zlib, withRootSupport ? false }:
```
Comment (jonringer) [inline pkgs/development/libraries/physics/yoda/default.nix]:
```suggestion
{ stdenv, fetchurl, fetchpatch, python, root, makeWrapper, zlib, withRootSupport ? false }:
```

File: pkgs/development/libraries/physics/yoda/default.nix [hunk 4]
```diff
@@ -9,14 +9,25 @@ stdenv.mkDerivation rec {
...
+  patches = [
+    # fixes "TypeError: expected bytes, str found" in writeYODA()
+    (fetchpatch {
+      url = "https://gitlab.com/hepcedar/yoda/commit/d2bbbe92912457f8a29b440cbfa0b39daf28ec34.diff";
+      sha256 = "1x60piswpxwak61r2sdclsc8pzi1fshpkjnxlyflsa1iap77vkq8";
+    })
+  ];
 
-  buildInputs = with python2Packages; [ python numpy matplotlib makeWrapper ]
+  nativeBuildInputs = with pythonPackages; [ cython ];
```
Comment (jonringer) [inline pkgs/development/libraries/physics/yoda/default.nix]:
```suggestion
  nativeBuildInputs = with python.pkgs; [ cython ];
```

File: pkgs/development/libraries/physics/yoda/default.nix [hunk 5]
```diff
@@ -9,14 +9,25 @@ stdenv.mkDerivation rec {
...
+  buildInputs = with pythonPackages; [ python numpy matplotlib makeWrapper ]
```
Comment (jonringer) [inline pkgs/development/libraries/physics/yoda/default.nix]:
```suggestion
  buildInputs = with python.pkgs; [ python numpy matplotlib makeWrapper ]
```

File: pkgs/top-level/python-packages.nix [hunk 6]
```diff
@@ -5634,6 +5634,10 @@ in {
 
   xxhash = callPackage ../development/python-modules/xxhash { };
 
+  yoda = toPythonModule (pkgs.yoda.override {
+    pythonPackages = self;
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]:
```suggestion
    inherit python;
```

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/top-level/all-packages.nix]: using my other suggestions:
```suggestion
    python = python3;
```

Comment (veprbl) [reply pkgs/top-level/all-packages.nix]: I will be switching `yoda` to use `python3` by default, but that, probably, can not be done until I get `rivet` version 3 working.

**Comment by veprbl** [comment]
@GrahamcOfBorg build python2Packages.yoda python3Packages.yoda

File: pkgs/development/libraries/physics/yoda/default.nix [hunk 7]
```diff
@@ -9,14 +9,26 @@ stdenv.mkDerivation rec {
...
+  buildInputs = [ python ]
+    ++ (with python.pkgs; [ numpy matplotlib ])
     ++ stdenv.lib.optional withRootSupport root;
   propagatedBuildInputs = [ zlib ];
 
   enableParallelBuilding = true;
 
+  postPatch = ''
+    touch pyext/yoda/*.{pyx,pxd}
+  '';
```
Comment (jonringer) [inline pkgs/development/libraries/physics/yoda/default.nix]: can you comment why this is necessary?

I'm assuming this is to fix some issue with cython?

Comment (veprbl) [reply pkgs/development/libraries/physics/yoda/default.nix]: This is needed to regenerate the cython C++ code  (remember, this is built with "make"), the one in the release tarball doesn't support python3.7.

Comment (jonringer) [reply pkgs/development/libraries/physics/yoda/default.nix]: ah, okay, force it to recompile. this makes sense

**Review by jonringer — APPROVED** [review]
diff looks better to me (improvement, would still like to see python3 eventually)
commit LGTM

```
https://github.com/NixOS/nixpkgs/pull/77642
10 package built:
fastnlo herwig mcgrid rivet yoda python37Packages.yoda python38Packages.yoda sherpa thepeg yoda-with-root
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build fastnlo herwig mcgrid rivet yoda python37Packages.yoda python38Packages.yoda sherpa thepeg yoda-with-root

**Comment by veprbl** [comment]
Let's merge this unless there are objections.

**Pull request merged**